File: leet-code/238. Product of Array Except Self/ProductOfArrayExceptSelf238_2.java
public class ProductOfArrayExceptSelf238_2 {
    public int[] productExceptSelf(int[] nums) {
        int[] answer = new int[nums.length];

        int total = 1;
        for (int i = 0; i < nums.length; i++) {
            answer[i] = total;
            total *= nums[i];
        }

        total = 1;
        for (int i = nums.length - 1; i > - 1; i--) {
            answer[i] *= total;
            total *= nums[i];
        }

        return answer;
    }
}

File: leet-code/238. Product of Array Except Self/ProductOfArrayExceptSelf238_2.kt
package com.example

class ProductOfArrayExceptSelf238_2 {
    fun productExceptSelf(nums: IntArray): IntArray {
        val answer = IntArray(nums.size)
        var p = 1
        for (i in nums.indices) {
            answer[i] = p
            p *= nums[i]
        }
        p = 1
        for (i in nums.size - 1 downTo 0) {
            answer[i] *= p
            p *= nums[i]
        }
        return answer
    }
}
